src/HaskellCI.hs

```diff
@@ -208,6 +212,36 @@ genTravisFromConfigs argv config prj versions' = do
                 , "# EOF"
                 ]
 
+-- | Adjust the generated Travis YAML output with patch files, if specified.
+-- We do this in a temporary file in case the user did not pass --output (as
+-- it would be awkward to patch the generated output otherwise).
+patchTravis
+    :: (MonadIO m, MonadMask m)
+    => Config -> [String] -> m [String]
+patchTravis cfg ls
+  | null patches = pure ls
+  | otherwise =
+      withSystemTempFile ".travis.yml.tmp" $ \fp h -> liftIO $ do
```

Comment:
what does `patch` do if patch application fails; it doesn't leave any conflict/orig/ etc files, does it?

Reply:
I haven't given much thought to how we should handle errors yet (which is one of the main reasons why I marked this as a WIP).

One option is to simply let the `patch` tool error out and cause `haskell-ci` to terminate. This could potentially leave some conflict files, but then again, it would leave them in the system's temporary directory, so perhaps this isn't so bad. Alternatively, we could try some more robust error-handling scheme.

What would you prefer?

Reply:
You would be the user, whatever is fine with you :)

Reply:
Hah, fair enough. I'll do some experimentation to see which choice seems better from an ergonomic perspective.

Reply:
Any outcome?

Reply:
Sorry for the delay—I haven't had a chance to work on this in a while.

Feel free to close this if you wish, as I can always reopen later if need be.

Reply:
No, it can be open, if you plan to continue to work on this

Reply:
Alright, I've had a chance to characterize the different ways this can fail, and what error messages you currently get in each situation:

1. The `.patch` file is missing:

   ```
   $ ../dist-newstyle/build/x86_64-linux/ghc-8.6.4/haskell-ci-0.3.20190413/x/haskell-ci/build/haskell-ci/haskell-ci regenerate
   *INFO* Generating Travis-CI config for testing for GHC versions: 7.8.1 7.8.2 7.8.3 7.8.4 7.10.1 7.10.2 7.10.3 8.0.1 8.0.2 8.2.1 8.2.2 8.4.1 8.4.2 8.4.3 8.4.4 8.6.1 8.6.2 8.6.3 8.6.4 8.6.5 8.8.1
   *ERROR* Cannot find cabal.project.travis-patch.patch
   ```
2. The `.patch` file contains garbage:

   ```
   $ ../dist-newstyle/build/x86_64-linux/ghc-8.6.4/haskell-ci-0.3.20190413/x/haskell-ci/build/haskell-ci/haskell-ci regenerate
   *INFO* Generating Travis-CI config for testing for GHC versions: 7.8.1 7.8.2 7.8.3 7.8.4 7.10.1 7.10.2 7.10.3 8.0.1 8.0.2 8.2.1 8.2.2 8.4.1 8.4.2 8.4.3 8.4.4 8.6.1 8.6.2 8.6.3 8.6.4 8.6.5 8.8.1
   patch: **** Only garbage was found in the patch input.
   haskell-ci: callProcess: patch "-i" "cabal.project.travis-patch.patch" "-s" "/tmp/.travis.yml3806-0.tmp" (exit 2): failed
   ```
3. The `.patch` file doesn't apply cleanly:

   ```
   $ ../dist-newstyle/build/x86_64-linux/ghc-8.6.4/haskell-ci-0.3.20190413/x/haskell-ci/build/haskell-ci/haskell-ci regenerate
   *INFO* Generating Travis-CI config for testing for GHC versions: 7.8.1 7.8.2 7.8.3 7.8.4 7.10.1 7.10.2 7.10.3 8.0.1 8.0.2 8.2.1 8.2.2 8.4.1 8.4.2 8.4.3 8.4.4 8.6.1 8.6.2 8.6.3 8.6.4 8.6.5 8.8.1
   1 out of 1 hunk FAILED -- saving rejects to file /tmp/.travis.yml4518-0.tmp.rej
   haskell-ci: callProcess: patch "-i" "cabal.project.travis-patch.patch" "-s" "/tmp/.travis.yml4518-0.tmp" (exit 1): failed
   ```

Some observations:

* Only (1) gives you an `*ERROR*`-prefixed message, due to the fact that I check for the absence of `.patch` files directly in `haskell-ci`. The other error messages simply barf out what `patch` tells you, so they're not formatted quite as nicely.
* Currently, all three forms of errors give you exit code 1, which is nice. However, the fact that (2) and (3) give you exit code 1 is probably just due to the fact that `patch` happens to return exit code 1, so I suppose there's nothing that guarantees this.

I don't know how to check for (2) or (3) directly in `haskell-ci`, so it seems likely that we'll have to rely on `patch` to error out. That being said, we could intercept the `patch` subprocess, check for its exit code, and format the errors a little nicer in the event that we receive a non-zero exit code. Do you think that would be worth it?

Reply:
I've just pushed a change so that we now use `readProcessWithExitCode` instead of `callProcess`. This means that we now have the following for scenarios (2) and (3):

2. ```
   $ ../dist-newstyle/build/x86_64-linux/ghc-8.6.4/haskell-ci-0.3.20190413/x/haskell-ci/build/haskell-ci/haskell-ci regenerate -o cabal.project.travis-patch.travis.yml 
   *INFO* Generating Travis-CI config for testing for GHC versions: 7.8.1 7.8.2 7.8.3 7.8.4 7.10.1 7.10.2 7.10.3 8.0.1 8.0.2 8.2.1 8.2.2 8.4.1 8.4.2 8.4.3 8.4.4 8.6.1 8.6.2 8.6.3 8.6.4 8.6.5 8.8.1
   *ERROR* patch returned exit code 2
   Stdout: 
   Stderr: patch: **** Only garbage was found in the patch input.
   
   ```
3. ```
   $ ../dist-newstyle/build/x86_64-linux/ghc-8.6.4/haskell-ci-0.3.20190413/x/haskell-ci/build/haskell-ci/haskell-ci regenerate -o cabal.project.travis-patch.travis.yml 
   *INFO* Generating Travis-CI config for testing for GHC versions: 7.8.1 7.8.2 7.8.3 7.8.4 7.10.1 7.10.2 7.10.3 8.0.1 8.0.2 8.2.1 8.2.2 8.4.1 8.4.2 8.4.3 8.4.4 8.6.1 8.6.2 8.6.3 8.6.4 8.6.5 8.8.1
   *ERROR* patch returned exit code 1
   Stdout: 1 out of 1 hunk FAILED -- saving rejects to file /tmp/.travis.yml32125-0.tmp.rej
   
   Stderr: 
   
   ```

I think this is probably as good as it's going to get, so I've removed the WIP label from this PR. Final thoughts?

Reply:
If travis is happy, I'll merge. Looks very good. Thanks!